// @mreal/digest-auth/build/main/lib/client-digest-auth.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const lodash_1 = require("lodash");
const constants_1 = require("./constants");
const incoming_digest_dto_1 = require("./dto/client/incoming-digest.dto");
const outgoing_digest_dto_1 = require("./dto/client/outgoing-digest.dto");
const outgoing_transform_digest_dto_1 = require("./dto/client/outgoing-transform-digest.dto");
const payload_dto_1 = require("./dto/client/payload.dto");
const cn_1 = require("./encryptions/cn");
const h1_1 = require("./encryptions/h1");
const h2_1 = require("./encryptions/h2");
const nonce_1 = require("./encryptions/nonce");
const response_1 = require("./encryptions/response");
const analyze_exception_1 = require("./exceptions/analyze-exception");
const base_exception_1 = require("./exceptions/base-exception");
const header_1 = require("./header");
const dto_1 = require("./utils/dto");
class ClientDigestAuth {
    static analyze(header, multipleAuthentication = false) {
        try {
            const challenges = header_1.Header.parse(header);
            const analyzeChallenges = challenges.map((challenge) => {
                if (challenge.scheme !== header_1.SCHEME_DIGEST)
                    return Object.assign({}, lodash_1.pick(challenge, ['scheme', 'raw']));
                const digest = dto_1.Dto.validate(incoming_digest_dto_1.IncomingDigestDto, challenge);
                return Object.assign({}, digest);
            });
            if (multipleAuthentication)
                return analyzeChallenges;
            const firstDigest = lodash_1.find(analyzeChallenges, { scheme: header_1.SCHEME_DIGEST });
            if (!firstDigest)
                throw new analyze_exception_1.AnalyzeException(analyze_exception_1.NOT_ALLOW_DIGEST);
            return firstDigest;
        }
        catch (e) {
            if (e instanceof base_exception_1.BaseException)
                throw e;
            throw new analyze_exception_1.AnalyzeException(analyze_exception_1.ANALYZE_CODE_UNEXPECTED);
        }
    }
    static generateUnprotected(serverDigest, username, password, payload) {
        const validPayload = dto_1.Dto.validate(payload_dto_1.PayloadUnprotectedDto, payload);
        const h1 = h1_1.HA1.generate(username, password, serverDigest.realm);
        const h2 = h2_1.HA2.generate(validPayload.method, validPayload.uri);
        const response = response_1.Response.generate(h1, serverDigest.nonce, h2);
        const digest = dto_1.Dto.validate(outgoing_digest_dto_1.OutgoingUnprotectedDigestDto, { nonce: serverDigest.nonce, realm: serverDigest.realm, response, username });
        return Object.assign({}, digest, { raw: header_1.Header.generate(Object.assign({}, dto_1.Dto.validate(outgoing_transform_digest_dto_1.OutgoingTransformUnprotectedDigestDto, digest))) });
    }
    static generateProtectionAuth(serverDigest, username, password, payload) {
        return this.generateQOP(Object.assign({}, serverDigest, { qop: constants_1.QOP_AUTH }), username, password, dto_1.Dto.validate(payload_dto_1.PayloadProtectionAuthDto, Object.assign({}, payload, { entryBody: '' })));
    }
    static generateProtectionAuthInt(serverDigest, username, password, payload) {
        return this.generateQOP(Object.assign({}, serverDigest, { qop: constants_1.QOP_AUTH_INT }), username, password, dto_1.Dto.validate(payload_dto_1.PayloadProtectionAuthIntDto, payload));
    }
    static generateQOP(serverDigest, username, password, payload) {
        const cnonce = nonce_1.Nonce.generate();
        const nc = cn_1.Cn.toHex(payload.counter);
        const algorithm = (payload.force_algorithm)
            ? payload.force_algorithm
            : serverDigest.algorithm || constants_1.ALGORITHM_MD5;
        const initialH1 = h1_1.HA1.generate(username, password, serverDigest.realm);
        const h1 = (algorithm === constants_1.ALGORITHM_MD5_SESS)
            ? h1_1.HA1.generateSess(initialH1, serverDigest.nonce, cnonce)
            : initialH1;
        const h2 = (serverDigest.qop === constants_1.QOP_AUTH_INT)
            ? h2_1.HA2.generateInt(payload.method, payload.uri, payload.entryBody)
            : h2_1.HA2.generate(payload.method, payload.uri);
        const response = response_1.Response.generateProtected(h1, serverDigest.nonce, h2, nc, cnonce, serverDigest.qop);
        const digest = dto_1.Dto.validate(outgoing_digest_dto_1.OutgoingProtectedDigestDto, {
            nonce: serverDigest.nonce,
            realm: serverDigest.realm,
            opaque: serverDigest.opaque,
            algorithm,
            qop: serverDigest.qop,
            uri: payload.uri,
            response,
            username,
            cnonce,
            nc
        });
        return Object.assign({}, digest, { raw: header_1.Header.generate(Object.assign({}, dto_1.Dto.validate(outgoing_transform_digest_dto_1.OutgoingTransformProtectedDigestDto, digest))) });
    }
}
exports.ClientDigestAuth = ClientDigestAuth;
